### ISSUE TITLE
Extract highs into highs_values, highs_indices

### ISSUE THREAD

**Issue by duycd2212** [AutodeskAILab/WaLa#13]
Hi. Thanks for your fantastic work

I am trying to reproduce your reconstruction results in encoder and decoder by using 256^3 TSDF and then extracting low, highs using your DWTForward3d function and get the shape low (1, 1, 7, 46, 46, 46) and highs [(1, 1, 7, 136, 136, 136), (1, 1, 7, 76, 76, 76), (1, 1, 7, 46, 46, 46)]. However, I have an issue extracting the highs into highs_values ([1, 46^3, 63]) and highs_indices ([1, 46^3, 63, 4]) to create a WaveletData object that can be encoded according to your code. 

I have tried using your extract_full_indices() function to get the highs_indices and then used it to get the highs_values, but it didn't work. The result of the reconstruction is shown below.
Can you give me instructions on how to extract highs_values and highs_indices from highs in your work, or are there any ways that you can encode the low and highs directly?

I'd be thankful if you give me an example data or preprocessing you are using in training or conditioning that can extract data["low"], data["high_indices"], data["high_values"]

Thank you 
![Screenshot from 2024-12-16 04-06-40](https://github.com/user-attachments/assets/f0b05279-8a56-4a93-b68e-a90edfdfeebc)

**Comment by kurtenkera** [comment]
Hi @duycd2212 have you had any luck with fixing this issue? I'm encountering the exact same problem! Perhaps we can work to resolve this together until the authors can help us out!

**Comment by duycd2212** [comment]
Hi @kurtenkera 
I figured it out after observing the output from the diffusion process. The main issue seems to be the preprocessing of the TSDF. You use the mesh_to_sdf library to convert .obj files to SDF, and then you clip the values exactly at (-0.1, 0.1). This range is probably what they use to train the encoder-decoder.

The extraction of `highs_indices` and `highs_values` is essentially the reverse of the step where `highs_recon` is produced from these values.

However, despite all of that, the output does not match the quality of those in the paper. I'm struggling to brainstorm additional ideas to replicate their preprocessing. I’m starting to think that the clipping + wavelet preprocessing and encoder-decoder might not be effective enough and could be causing a loss of information.
![Screenshot from 2025-01-09 14-58-36](https://github.com/user-attachments/assets/5aab3125-e794-47d5-992f-4dd933b5d964)

**Comment by duycd2212** [comment]
Thank you, @kurtenkera. Your information is extremely helpful to me. After reviewing the GitHub repository on _Neural Wavelet-domain Diffusion for 3D Shape Generation_, here is my current preprocessing pipeline:
.obj file -> Flip -> Convert to manifold watertight object -> TSDF (scale_ratio=0.9 + 0.1 clipping) -> Wavelet

The result seems to match the paper, but it still does not work well with complex objects.
![Screenshot from 2025-01-10 23-54-02](https://github.com/user-attachments/assets/5ffe040a-71ca-453b-ba8f-f8bb586209f4)

Regarding your question about high_indices and high_values: Yes, I use reverse engineering to derive highs_values from highs and highs_indices. I obtained the highs_indices from the extract_full_indices() function (They filtered low and highs as outlined in the paper), followed by some postprocessing:
`        highs_indices = extract_full_indices(device = 'cuda', max_depth=self.args.max_depth, shape_list=self.dwt_sparse_composer.shape_list)
        output_dim = 8**self.args.max_training_level - 1
        highs_indices = highs_indices[:, -output_dim:]
        padded_highs_indices = padding_high_indices(batch_size=1, highs_indices=highs_indices)   `

I input this and highs into batch_extract_highs_values(), which reverses extract_highs_from_values(). I implemented the code myself and am happy to share it with you. Some of the code is hard-coded based on the configs.

`def batch_extract_highs_values(
    output_stage=2,
    max_depth=3,
    shape_list = [[256, 256, 256], [136, 136, 136], [76, 76, 76], [46, 46, 46]],
    device = 'cuda',
    batch_size = 1,
    highs_indices=None,
    highs_recon=None,
):
    cnt = 0
    highs_values_list = []
    for idx in range(max_depth):
        current_stage = (max_depth - 1) - idx # 2 1 0

        padding_size = (
            shape_list[-1][0] * (2**current_stage) - shape_list[idx + 1][0]
        ) // 2 # 8  0

        if current_stage < output_stage:
            assert highs_indices.size(0) == batch_size

            indices_len = 7 * ((2**current_stage) ** 3) #  56  7
            high_indices_idx = highs_indices[:, :, cnt : cnt + indices_len] #[1,:,:,0:56] [1,:,:,56:63]
            high_indices_idx = high_indices_idx.reshape((-1, 5)).long() 

            high_new = highs_recon[idx]
            high_new = torch.nn.functional.pad(high_new, (padding_size, padding_size, padding_size, padding_size, padding_size, padding_size))

            high_values_idx = high_new[
                high_indices_idx[:, 0],
                0,
                high_indices_idx[:, 1],
                high_indices_idx[:, 2],
                high_indices_idx[:, 3],
                high_indices_idx[:, 4],
            ]
            high_values_idx = high_values_idx.view(batch_size, -1, indices_len)
            highs_values_list.append(high_values_idx)
            cnt += indices_len
    highs_values = torch.cat(highs_values_list, dim=-1)
    return highs_values`

Thank you once again. If you have any questions or feedback, I would be happy to hear from you.

**Comment by duycd2212** [comment]
Thank you, @kurtenkera. Your information is extremely helpful to me. After reviewing the GitHub repository on Neural Wavelet-domain Diffusion for 3D Shape Generation, here is my current preprocessing pipeline:
.obj file -> Flip -> Convert to manifold watertight object -> TSDF (scale_ratio=0.9 + 0.1 clipping) -> Wavelet

The result seems to match the paper, but it still does not work well with complex objects.
Screenshot from 2025-01-10 23-54-02

Regarding your question about high_indices and high_values: Yes, I use reverse engineering to derive highs_values from highs and highs_indices. I obtained the highs_indices from the extract_full_indices() function (They filtered low and highs as outlined in the paper), followed by some postprocessing:

```
        highs_indices = extract_full_indices(device = 'cuda', max_depth=self.args.max_depth, shape_list=self.dwt_sparse_composer.shape_list)
        output_dim = 8**self.args.max_training_level - 1
        highs_indices = highs_indices[:, -output_dim:]
        padded_highs_indices = padding_high_indices(batch_size=1, highs_indices=highs_indices)      
```

I input this and highs into batch_extract_highs_values(), which reverses extract_highs_from_values(). I implemented the code myself and am happy to share it with you. Some of the code is hard-coded based on the configs.

```
def batch_extract_highs_values(
output_stage=2,
max_depth=3,
shape_list = [[256, 256, 256], [136, 136, 136], [76, 76, 76], [46, 46, 46]],
device = 'cuda',
batch_size = 1,
highs_indices=None,
highs_recon=None,
):
cnt = 0
highs_values_list = []
for idx in range(max_depth):
current_stage = (max_depth - 1) - idx # 2 1 0

    padding_size = (
        shape_list[-1][0] * (2**current_stage) - shape_list[idx + 1][0]
    ) // 2 # 8  0

    if current_stage < output_stage:
        assert highs_indices.size(0) == batch_size

        indices_len = 7 * ((2**current_stage) ** 3) #  56  7
        high_indices_idx = highs_indices[:, :, cnt : cnt + indices_len] #[1,:,:,0:56] [1,:,:,56:63]
        high_indices_idx = high_indices_idx.reshape((-1, 5)).long() 

        high_new = highs_recon[idx]
        high_new = torch.nn.functional.pad(high_new, (padding_size, padding_size, padding_size, padding_size, padding_size, padding_size))

        high_values_idx = high_new[
            high_indices_idx[:, 0],
            0,
            high_indices_idx[:, 1],
            high_indices_idx[:, 2],
            high_indices_idx[:, 3],
            high_indices_idx[:, 4],
        ]
        high_values_idx = high_values_idx.view(batch_size, -1, indices_len)
        highs_values_list.append(high_values_idx)
        cnt += indices_len
highs_values = torch.cat(highs_values_list, dim=-1)
return highs_values
```
Thank you once again. If you have any questions or feedback, I would be happy to hear from you.

**Comment by duycd2212** [comment]
Thank you, @kurtenkera. Your information is extremely helpful to me. After reviewing the GitHub repository on Neural Wavelet-domain Diffusion for 3D Shape Generation, here is my current preprocessing pipeline:
.obj file -> Flip -> Convert to manifold watertight object -> TSDF (scale_ratio=0.9 + 0.1 clipping) -> Wavelet

The result seems to match the paper, but it still does not work well with complex objects.
![Screenshot from 2025-01-10 23-54-02](https://github.com/user-attachments/assets/4aa93ba1-8329-4c64-9a32-20c4a2e93f85)

Regarding your question about high_indices and high_values: Yes, I use reverse engineering to derive highs_values from highs and highs_indices. I obtained the highs_indices from the extract_full_indices() function (They filtered low and highs as outlined in the paper), followed by some postprocessing.

I implemented the code myself and am happy to share it with you. Some of the code is hard-coded based on the configs.

```
def batch_extract_highs_values(
output_stage=2,
max_depth=3,
shape_list = [[256, 256, 256], [136, 136, 136], [76, 76, 76], [46, 46, 46]],
device = 'cuda',
batch_size = 1,
highs_indices=None,
highs_recon=None,
):
cnt = 0
highs_values_list = []
for idx in range(max_depth):
current_stage = (max_depth - 1) - idx # 2 1 0

    padding_size = (
        shape_list[-1][0] * (2**current_stage) - shape_list[idx + 1][0]
    ) // 2 # 8  0

    if current_stage < output_stage:
        assert highs_indices.size(0) == batch_size

        indices_len = 7 * ((2**current_stage) ** 3) #  56  7
        high_indices_idx = highs_indices[:, :, cnt : cnt + indices_len] #[1,:,:,0:56] [1,:,:,56:63]
        high_indices_idx = high_indices_idx.reshape((-1, 5)).long() 

        high_new = highs_recon[idx]
        high_new = torch.nn.functional.pad(high_new, (padding_size, padding_size, padding_size, padding_size, padding_size, padding_size))

        high_values_idx = high_new[
            high_indices_idx[:, 0],
            0,
            high_indices_idx[:, 1],
            high_indices_idx[:, 2],
            high_indices_idx[:, 3],
            high_indices_idx[:, 4],
        ]
        high_values_idx = high_values_idx.view(batch_size, -1, indices_len)
        highs_values_list.append(high_values_idx)
        cnt += indices_len
highs_values = torch.cat(highs_values_list, dim=-1)
return highs_values
```
Thank you once again. If you have any questions or feedback, I would be happy to hear from you.

**Comment by kurtenkera** [comment]
Thank you, @duycd2212, this is helpful! 

Did you have to alter `batch_pad` or `self.highs_indices` in the `convert_wavelet_volume()` method belonging to the `WaveletData` class? 

When I instantiate a `WaveletData` object` and call the `convert_wavelet_volume()` function, a bug occurs due to the following:

```python
    ### extend batch_size to the indices
    batch_pad = (
        torch.arange(batch_size, device=self.highs_indices.device)
        .unsqueeze(1)
        .repeat((1, self.highs_indices.size(1)))
        .unsqueeze(2)
        .long()
    )  # B * P * 1
    
    high_indices_filled = torch.cat(
        (batch_pad, self.highs_indices), dim=2
    )  # B * P * 4
```

The error is due to the fact that `batch_pad` has shape [1, 46^3, 1] and `self.highs_indices` has shape [1, 46^3, 63, 4], and so the two tensors cannot be concatenated in `high_indices_filled`?

**Comment by kurtenkera** [comment]
Hi @duycd2212, just following up about my questions above?

**Comment by kurtenkera** [comment]
I haven't had any success, but I stopped working on this probably around the end of January as I'm working on a different project. I'll return to this after I finish my current project, and I'll let you know when I succeed (I'm confident I'll get it to work)!

**Comment by NolimitDougie** [comment]
> I haven't had any success, but I stopped working on this probably around the end of January as I'm working on a different project. I'll return to this after I finish my current project, and I'll let you know when I succeed (I'm confident I'll get it to work)!

Okay I got the high indices but still trying to extract the  high values from the highs. I'm using there     extract_highs_from_values() function to try but not getting any success on it
